### PR TITLE
[lldb] Add support for disabling frame recognizers

### PR DESCRIPTION
**Pull request by vogelsgesang**
Sometimes you only want to temporarily disable a frame recognizer instead of deleting it. In particular, when dealing with one of the builtin frame recognizers, which cannot be restored after deletion.

To be able to write test cases for this functionality, I also changed `lldb/test/API/commands/frame/recognizer` to use normal C instead of Objective-C

### CODE REVIEW

File: lldb/source/Commands/CommandObjectFrame.cpp [hunk 1]
```diff
@@ -957,53 +962,43 @@ PrintRecognizerDetails(Stream &strm, const std::string &name,
...
-      }
-
-      GetTarget().GetFrameRecognizerManager().RemoveAllRecognizers();
-      result.SetStatus(eReturnStatusSuccessFinishResult);
-      return;
-    }
+  virtual void DoExecuteWithId(CommandReturnObject &result, uint32_t recognizer_id) = 0;
 
+  void DoExecute(Args &command, CommandReturnObject &result) override {
     if (command.GetArgumentCount() != 1) {
```
Comment (jimingham) [inline lldb/source/Commands/CommandObjectFrame.cpp:1002]: You didn't add this part, so feel free to ignore this comment, but if you pass `eArgRepeatOptional` to defaulted second parameter to AddSimpleArgumentList, you shouldn't need to do this check here.

**Comment by jimingham** [comment]
This looks good to me so far, but of course it needs tests.

**Comment by vogelsgesang** [comment]
This is ready for review now.

While adding the test cases, I had to migrate it from Objective-C to C, since I was unable to run it on my Linux machine otherwise

File: lldb/test/API/commands/frame/recognizer/TestFrameRecognizer.py [hunk 2]
```diff
@@ -395,6 +389,138 @@ def test_frame_recognizer_not_only_first_instruction(self):
...
+                "0: recognizer.MyFrameRecognizer, module a.out, demangled symbol foo"
+            ],
+        )
+
+        self.expect(
+            "frame recognizer info 0",
+            substrs=["frame 0 is recognized by recognizer.MyFrameRecognizer"],
+        )
+
+    def test_frame_recognizer_disable(self):
```
Comment (Michael137) [inline lldb/test/API/commands/frame/recognizer/TestFrameRecognizer.py:448]: FYI, duplicate test-names will result in only the last declared one being run (see https://github.com/llvm/llvm-project/pull/97043)

File: lldb/test/API/commands/frame/recognizer/TestFrameRecognizer.py [hunk 3]
```diff
@@ -395,6 +389,138 @@ def test_frame_recognizer_not_only_first_instruction(self):
...
+        self.build()
+        exe = self.getBuildArtifact("a.out")
```
Comment (Michael137) [inline lldb/test/API/commands/frame/recognizer/TestFrameRecognizer.py:450]: What newly added functionality is this test-case actually testing? Couldn't quite tell

File: lldb/test/API/commands/frame/recognizer/categories [hunk 4]
```diff
@@ -1 +0,0 @@
-objc
```
Comment (Michael137) [inline lldb/test/API/commands/frame/recognizer/categories:1]: Hmm are we sure we aren't losing test coverage by converting this to plain C? I don't think it'd be a huge deal if we just added an entirely new test as part of this PR

Comment (Michael137) [reply lldb/test/API/commands/frame/recognizer/categories:1]: Maybe with some Makefile trickery you could add a separate main.c in this directory (in addition to `main.m`). But that's probably not worth the trouble

**Review by Michael137 — APPROVED** [review]
Modulo test comments, LGTM

File: lldb/test/API/commands/frame/recognizer/TestFrameRecognizer.py (hunk 3 above)
Comment (vogelsgesang) [reply lldb/test/API/commands/frame/recognizer/TestFrameRecognizer.py]: copy-paste mistake. Removed, and thereby also solved the issue with duplicated test case names

File: lldb/test/API/commands/frame/recognizer/categories (hunk 4 above)
Comment (vogelsgesang) [reply lldb/test/API/commands/frame/recognizer/categories:1]: The intended test coverage of this test case seems to be to test the frame recognizers as indicated by the file path.
Making making it run also on non-Mac platforms, we are hence gaining test coverage on other systems.

There is no comment I could find which states that frame recognizers work differently for Objective-C. As such, I would expect any test coverage for Objective-C to be incidental, and would expect the Objective-C support to be tested in-depth by some other test cases.

I am used to the guideline "It's ok to remove incidental test coverage. If the test coverage was worthwhile, there should have been a comment added as part of adding the test" from other projects. But that's based on my experience with how other projects treat their test suites. Not sure if this also applies to LLDB?

Comment (Michael137) [reply lldb/test/API/commands/frame/recognizer/categories:1]: Yea that's fair, there doesn't seem to be anything objc specific here, so seems ok to turn it into C and enable it on all platforms.

The original change that introduced these did mention Objective-C as one of the motivators: https://reviews.llvm.org/D44603. And there seems to have been changes made to `AppleObjCStackFrameRecognizer.cpp` in early version of the revision (though Phabricator doesn't allow me to view those).

So overall I'm ok with this

**Review by Michael137 — APPROVED** [review]

File: lldb/test/API/commands/frame/recognizer/categories (hunk 4 above)
Comment (jimingham) [reply lldb/test/API/commands/frame/recognizer/categories:1]: The early motivation for stack frame recognizers was to get the "current exception object" from the objc_exception_throw routine.  But this patch didn't add that recognizer, and when it was added it had its own test.  There's nothing about this test that's ObjC specific, and the ObjC specific bits that motivated making the test file a .m file were actually tested in a separate commit.  So I'm also fine with changing this test to a .c.

Comment (vogelsgesang) [reply lldb/test/API/commands/frame/recognizer/categories:1]: Thanks for confirming!

**Comment by medismailben** [comment]
@vogelsgesang Nice!

**Pull request merged**